core/trino-main/src/main/java/io/trino/execution/SqlTask.java

```diff
@@ -167,19 +171,22 @@ private void initialize(Consumer<SqlTask> onDone, CounterStat failedTasks)
             }
 
             // store final task info
-            while (true) {
+            taskHolderLock.lock();
```

Comment:
nit: consider `synchronized(this)` and reorder the `taskExecution.start();` as discussed offline (unless we missed something that can break the flow)